src/transformers/models/dac/modeling_dac.py

```diff
@@ -0,0 +1,616 @@
+        super().__init__()
+        self.codebook_size = codebook_size
+        self.codebook_dim = codebook_dim
+
+        self.in_proj = WNConv1d(input_dim, codebook_dim, kernel_size=1)
+        self.out_proj = WNConv1d(codebook_dim, input_dim, kernel_size=1)
+        self.codebook = nn.Embedding(codebook_size, codebook_dim)
+
+    def forward(self, input):
+        """
+        Quantizes the input tensor using a fixed codebook and returns the corresponding codebook vectors.
+
+        Args:
+            input (`torch.Tensor` of shape `(batch_size, dimension, time_steps)`): 
+                Input tensor.
+
+        Returns:
+            quantized_representation (`torch.Tensor`of shape `(batch_size, dimension, time_steps)`). 
+                Quantized continuous representation of input, .
+            commitment_loss : torch.Tensor
+                Commitment loss to train encoder to predict vectors closer to codebook entries, shape `(1)`.
+            codebook_loss : torch.Tensor
+                Codebook loss to update the codebook, shape `(1)`.
+            codebook_indices (`torch.Tensor` of shape `(batch_size, time_steps)`): 
+                Codebook indices for each codebook, quantized discrete representation of input,
+            projected_latents : torch.Tensor of shape `(batch_size, num_codebooks * dimension, time_steps)`
+                Projected latents (continuous representation of input before quantization),
+        """
+
+        projected_latents = self.in_proj(input)  
+        quantized_representation, codebook_indices = self.decode_latents(projected_latents)
+
+        commitment_loss = F.mse_loss(projected_latents, quantized_representation.detach(), reduction="none").mean([1, 2])
+        codebook_loss = F.mse_loss(quantized_representation, projected_latents.detach(), reduction="none").mean([1, 2])
+
+        quantized_representation = projected_latents + (quantized_representation - projected_latents).detach()  # noop in forward pass, straight-through gradient estimator in backward pass
+
+        quantized_representation = self.out_proj(quantized_representation)
+
+        return quantized_representation, commitment_loss, codebook_loss, codebook_indices, projected_latents
+
+    def embed_code(self, embed_id):
+        """Embeds a code using the codebook weights.
+        """
+        return F.embedding(embed_id, self.codebook.weight)
+
+    def decode_code(self, embed_id):
+        """Decodes embedded codes by embedding them and then transposing the dimensions.
+        """
+        return self.embed_code(embed_id).transpose(1, 2)
+
+    def decode_latents(self, latents):
+        encodings = rearrange(latents, "b d t -> (b t) d")
+        codebook = self.codebook.weight  # codebook: (N x D)
+
+        # L2 normalize encodings and codebook (ViT-VQGAN)
+        encodings = F.normalize(encodings)
+        codebook = F.normalize(codebook)
+
+        # Compute euclidean distance with codebook
+        dist = (
```

Comment:
This is a bit cluttered. Could we clean it with something like:
```suggestion
        l2_norm = encodings.pow(2).sum(1, keepdim=True)
        dist = (l2_norm - 2 * encodings @ codebook.t()) + l2_norm.t()
```